datalad/distribution/dataset.py

```diff
@@ -187,22 +189,61 @@ def path(self):
     @property
     def repo(self):
         """Get an instance of the version control system/repo for this dataset,
-        or None if there is none yet.
+        or None if there is none yet (or none anymore).
 
-        If creating an instance of GitRepo is guaranteed to be really cheap
-        this could also serve as a test whether a repo is present.
+        If testing the validity of an instance of GitRepo is guaranteed to be
+        really cheap this could also serve as a test whether a repo is present.
+
+        Note, that this property is evaluated every time it is used. If used
+        multiple times within a function it's probably a good idea to store its
+        value in a local variable and use this variable instead.
 
         Returns
         -------
-        GitRepo
+        GitRepo or AnnexRepo
         """
 
-        # Note: lazy loading was disabled, since this is provided by the
-        # flyweight pattern already and a possible invalidation of an existing
-        # instance has to be done therein.
-        # TODO: Still this is somewhat problematic. We can't invalidate strong
-        # references
-
+        # If we already got a *Repo instance, check whether it's still valid;
+        # Note, that this basically does part of the testing that would
+        # (implicitly) be done in the loop below again. So, there's still
+        # potential to speed up when we actually need to get a new instance
+        # (or none). But it's still faster for the vast majority of cases.
+        #
+        # TODO: Dig deeper into it and melt with new instance guessing. This
+        # should also involve to reduce redundancy of testing such things from
+        # within Flyweight.__call__, AnnexRepo.__init__ and GitRepo.__init__!
+        #
+        # Also note, that this could be forged into a single big condition, but
+        # that is hard to read and we should be well aware of the actual
+        # criteria here:
+        if self._repo is not None and realpath(self.path) == self._repo.path:
+            # we got a repo and path references still match
+            if isinstance(self._repo, AnnexRepo):
+                # it's supposed to be an annex
+                if self._repo is AnnexRepo._unique_instances.get(
+                        self._repo.path, None) and \
+                        AnnexRepo.is_valid_repo(self._repo.path,
+                                                allow_noninitialized=True):
+                    # it's still the object registered as flyweight and it's a
+                    # valid annex repo
+                    return self._repo
+            elif isinstance(self._repo, GitRepo):
+                # it's supposed to be a plain git
+                if self._repo is GitRepo._unique_instances.get(
+                        self._repo.path, None) and \
+                        GitRepo.is_valid_repo(self._repo.path) and not \
```

Comment:
the same here -- when could it already be `GitRepo` but no longer a valid git repo?

Reply:
Remove/uninstall. Or something went really wrong and you killed .git. The point is: Just because something was there, it doesn't mean it's there forever.

Reply:
well, harddrive might disappear too!  Should we then constantly sample for the `/` to still be there? my point is that all these operations cost time, and since if "something went really wrong and you killed .git" you are then screwed anyways, it would be nice to avoid wasting that time

Reply:
> if "something went really wrong and you killed .git" you are then screwed anyways

Sure, but you still want to fail reasonably in that case instead of pretending everything was in order, don't you? And apart from that, as I pointed out before, even a simple uninstall would cause `self.repo` to be invalid if we wouldn't update that property. Have a look at the tests I added.
Plus: As far as `self.repo` is concerned, we do it in master as well, but even MORE expensive! This is faster than before! (And yes, it could be even faster and I'll do it) The entire point of this PR is, that we did it only for `repo` and thus decoupling `config` and `id`, since we didn't do it for them. In master we simply end up in an inconsistent state.
I can't see how this discussion would prevent merging. If we want to do it differently altogether (and change the already existing tests that would fail ATM if this was done differently), we can have that discussion, but then we need to come up with (and agree on) a solution that at the very least results in a consistent state.

In case I somewhat misread your objection and it's really just about figuring out, what exact conditions really need checking: Yes, there's room for improvement. I just really want to do it in a dedicated PR, since this involves tuning of what is done in flyweight and in the *Repo constructors and in the `is_valid_repo` methods (particularly in case of annex). I'm happy to discuss the details of that, since we do redundant things indeed. But I need this bug fixed first and further optimize the validation second.